### DISCUSSION TITLE
Can you change the loss function of the PySRRegressor object to predict certain functions?

### DISCUSSION THREAD

**Discussion by unary-code** [MilesCranmer/PySR#356] [Q&A]
Deleted because problem is resolved

**Comment by MilesCranmer** [comment]
This might be helpful: https://astroautomata.com/PySR/examples/#9-custom-objectives

One tricky thing to note is that there isn’t really a way to normalize the symbolic expression.

**Reply by unary-code** [reply]
Hi Mr. Cranmer,

I'd like to find a way to normalize the symbolic expression.
Let's assume that our original dataset comes from a true probability distribution of a finite-sized domain, so for example our original dataset comes from the uniform distribution over [0, 1).

Julia has a method to take the definite integral of a function over a finite-sized domain.
Specifically, integrate(exp(-x), (x, 0, oo)) gives you the value of the definite integral of e^(-x) from x=0 to x=+infinity.

I know the objective function has 3 arguments (tree, dataset, and options).
"tree" holds the tree which represents the PySRRegressor model's guessed math expression.
I could find a way to obtain the string-representation tree_as_a_str of the math expression represented by the tree "tree", for example "exp(x - 5)" would be the string-representation of the tree (see the attached image of the tree).
Then, I could do "integrate(tree_as_a_str, (x, 0, 1))" in order to get the value of the definite integral of the guessed math expression "tree_as_a_str" from x=0 to x=1.

Or, alternatively, if it is hard for me to generate the string representation of the math expression represented by "tree", then I could use a function "eval_tree_array(tree, convert_float_to_dataset(x), options)" where convert_float_to_dataset(x) will take a single floating number x and convert it to a Dataset type of size one element.
Then I could do:  "integrate(eval_tree_array(tree, convert_float_to_dataset(x), options) (x, 0, 1))".

Do these seem to you like viable approaches that would work?
Are there more efficient ways to normalize the distribution with your PySR package?

**Comment by unary-code** [comment]
Hi Mr. Cranmer,

Thank you for responding!
I am a little confused about how to write a custom objective function.
If I wanted the loss function to be the Mean-Squared-Error between the value of the math expression called "tree", when evaluated across all N data points in the dataset, and the true value of y for those N data points in the dataset, then could my custom objective function be the following string:

```python
objective = """
function my_custom_objective(tree, dataset::Dataset{T,L}, options) where {T,L}
    # Evaluate numerator:
    Tree_prediction, flag = eval_tree_array(tree, dataset.X, options)
    !flag && return L(Inf)

    diffs = prediction .- dataset.y

    return sum(diffs .^ 2) / length(diffs)
end
""" 
```



In addition, I have a clarification question about the provided sample code for a custom objective function on https://astroautomata.com/PySR/examples/#9-custom-objectives
What exactly is this custom objective function doing?
My idea of what the website's provided objective function is doing: It is making sure that the neural network will find two polynomial expressions P and Q in terms of x, such that the value of y in the dataset is approximated by (close to) the value of P(x) / Q(x) for each data point.
Using this objective function essentially forces the meaning of the topmost root node of the tree to be a "/" divide-by operator.
What would the advantage of this particular objective function be? If you know the dataset's y variable can be described by a function of x divided by another function of x, then you would use this particular objective function (instead of the general objective function that most people use for symbolic regression)?
If the true math expression was a division of a polynomial of x by another polynomial of x (let's call this Case "dataset's y is a polynomial division"), would the general objective function that most people use for symbolic regression (the MSE between the value of the neural network's outputted math expression, when evaluated across all N data points in the dataset, and the true value of y for those N data points in the dataset) hypothetically be able to find the true math expression? Or under Case "dataset's y is a polynomial division" maybe would it take more time for the general objective function to find the true math expression than if you used that particular objective function, because the general objective function would probably search using non-"/" nodes as the top-most root node of the tree, which would be a waste of time for this dataset (the function whose code is in https://astroautomata.com/PySR/examples/#9-custom-objectives)?

Thanks, appreciate your time very much.

**Reply by MilesCranmer** [reply]
> If I wanted the loss function to be the Mean-Squared-Error between the value of the math expression called "tree", when evaluated across all N data points in the dataset, and the true value of y for those N data points in the dataset, then could my custom objective function be the following string:


Yes that's correct. There are also some built-in loss functions if you are doing something simple and elementwise: https://astroautomata.com/PySR/api/#the-objective (e.g., MSE, mean absolute-error, etc).

> In addition, I have a clarification question about the provided sample code for a custom objective function on https://astroautomata.com/PySR/examples/#9-custom-objectives
> What exactly is this custom objective function doing?
> My idea of what the website's provided objective function is doing: It is making sure that the neural network will find two polynomial expressions P and Q in terms of x, such that the value of y in the dataset is approximated by (close to) the value of P(x) / Q(x) for each data point.
> Using this objective function essentially forces the meaning of the topmost root node of the tree to be a "/" divide-by operator.
> What would the advantage of this particular objective function be?

That specific example may not apply to your problem because it is trying to enforce a particular symbolic form (that the equation is a fraction of polynomials). The purpose of this example is just to demo that the objective is very flexible. Anything you would want to optimize, you can probably put it there.

Cheers,
Miles

**Comment by unary-code** [comment]
Hi,
Thanks for your previous responses: they help me understand how to make my own loss function and set that as the objective function of the PySRRegressor object.

However, I am having trouble finding a loss funciton that will result in the PySRRegressor finding a accurate probability distribution that is accurate to the true probability distribution.
I have tried:
objective = """
function eval_loss(tree, dataset::Dataset{T,L}, options)::L where {T,L}
    # Evaluate numerator:
    prediction, flag = eval_tree_array(tree, dataset.X, options)
    !flag && return L(Inf)


    return -1 * sum(log.(prediction.*prediction)) / (2*length(prediction))
end
"""
but even when using this particular loss function, which I expected would result in an accurate predicted probability distribution, the PySRRegressor object does not accurately find a probability distribution.

Any advice?
Any papers or terms I could look up so I could learn some ideas for what to put as the loss function?


I have learned about how a GAN can generate samples which collectively approximate the unknown true probability distribution, but I don't think your PySR uses a GAN. Could I perhaps use the loss method of a GAN or another method for density estimation?
